### PR TITLE
Font face permissions

### PR DESCRIPTION
**Pull request by peterwilsoncc**
## What?

Modifies the font face permissions to ensure the file system is writable for the installation and deletion of fonts.

The font library's JavaScript is modified to use these permission to determine whether to allow:

* installation of google fonts
* upload of font files
* deletion of fonts

See #55280

## Why?

Currently the `DISALLOW_FILE_MODS` constant is not checked as part of the font faces API allowing users to upload fonts when the fonts should not be disabled.

As this constant is used to indicate an immutable file system, the constant needs to be considered to allow for a good user experience uploading fonts. See https://github.com/WordPress/wporg-main-2022/issues/388, https://github.com/WordPress/gutenberg/issues/55063

## How?

* Induces the meta capabilities: `install_font_faces`, `delete_font_faces`.
* Checks `[core data store].canUser( 'create', 'font-families' );` to determine if a user can upload, removes the UI if not
* Checks `[core data store].canUser( 'delete', 'font-families' );` to determine if a user can delete, removes the UI if not

> [!NOTE]
>
> Please test running this branch on **WordPress 6.4**.
>
> The file the changes have been made in are not loaded on the current version of WordPress trunk.

## Testing Instructions

1. Install a font either via file upload or Google Fonts.
2. Add `define( 'DISALLOW_FILE_MODS', true );` to your wp-config file.
3. Open Site Editor ... Styles > Typography > Manage Fonts
4. Ensure the upload and Install fonts tabs are not shown
5. Click the font you installed earlier
6. Check the box next to the font
7. Ensure the delete button is not available
8. Remove the constant added earlier
9. Ensure it's possible to delete your font.

### CODE REVIEW

File: packages/edit-site/src/components/global-styles/font-library-modal/index.js [hunk 1]
```diff
@@ -45,9 +50,15 @@ function FontLibraryModal( {
 } ) {
 	const { collections, setNotice } = useContext( FontLibraryContext );
 
+	const canUserCreate = useSelect( ( select ) => {
+		const { canUser } = select( coreStore );
+		return canUser( 'create', 'font-families' );
+	} );
```
Comment (peterwilsoncc) [inline packages/edit-site/src/components/global-styles/font-library-modal/index.js:56]: This should really be checking the permission of the `wp_font_face` post type but I am not sure how to do that. Putting `font-families/(?P<font_family_id>[\d]+)/font-faces` in didn't seem to work. Can I just use the post type?

I'm also not too clear on the difference between `canUser` and `useResourcePermissions` 

Same applies to the deletion capability check in `installed-fonts.js`

Comment (youknowriad) [reply packages/edit-site/src/components/global-styles/font-library-modal/index.js:56]: Think of `canUser( requestedAction, resource, id )` as a shortcut to check the "allow headers" for `wp/v2/:resource/:id` (id being optional). so for `create`, it's going to check whether the `POST`  method is  allowed for that resource (using an OPTIONS request).

I don't think we support nested routes like that yet. We could potentially make a change to `core-data` to support it, but do you know which REST API could give us this information? 

useResourcePermissions is just a shortcut to retrieve multiple permissions for a given resource. (It's a shortcut towards multiple canUser calls)

Comment (swissspidy) [reply packages/edit-site/src/components/global-styles/font-library-modal/index.js:56]: `canUser( 'create', 'font-families/123/font-faces' )` would probably work, but you'd need the font family ID.

**Comment by matiasbenedetto** [comment]
Hi @peterwilsoncc :wave: , thanks for working on this.

> 1. Install a font either via file upload or Google Fonts.
> 2. Add `define( 'DISALLOW_FILE_MODS', true );` to your wp-config file.
> 3. Open Site Editor ... Styles > Typography > Manage Fonts
> 4. Ensure the upload and Install fonts tabs are not shown
> 5. Click the font you installed earlier
> 6. Check the box next to the font
> 7. Ensure the delete button is not available
> 8. Remove the constant added earlier
> 9. Ensure it's possible to delete your font.

I tested this PR following the instructions, and it works as described.


If understood correctly [from your comment](https://github.com/WordPress/gutenberg/issues/55280#issuecomment-1942698730) defining `define( 'DISALLOW_FILE_MODS,' true );` disallow file writing on all the folders except the uploads ones. I have a question: What behavior would be expected if the user customized the fonts folder inside the `uploads` folder [as this extender did](https://github.com/WordPress/gutenberg/issues/58696)? In theory, the user would allowed to write fonts because, using that config, they will be written inside the uploads folder, but, running this PR, the tabs with the UI to install the fonts are not available.

**Comment by swissspidy** [comment]
It doesn‘t matter where the fonts folder is located. The capability check prevents adding fonts if file mods are disallowed.

**Comment by creativecoder** [comment]
There's some nuance to how fonts work that seems important to consider here:

- `wp_font_face` posts contain a url in the src property of the font face JSON that point to the font file
- The referenced font file url can be local (font file is uploaded to the site) or remote (font is hosted on a CDN)

So I think that permission to create a wp_font_face post should stay separate from uploading a font file. If `DISALLOW_FILE_MODS` is true, I don't think that should prevent creating all font faces (which can reference a remotely hosted file), only the actual font file upload.

Perhaps we could hand this by

- Not showing the font upload UI in the Font Library when `DISALLOW_FILE_MODS` is true
- Making sure that the POST font-faces endpoint disallows font file uploads when `DISALLOW_FILE_MODS` is true

Additionally, there's a discussion in [this issue](https://github.com/WordPress/gutenberg/issues/58229) about how to signal when to upload a font file to the site (the current behavior with the Google Fonts library) vs using the remotely hosted font file. I'll add a comment there, I think that `DISALLOW_FILE_MODS` needs to be considered.

File: packages/edit-site/src/components/global-styles/font-library-modal/index.js (hunk 1 above)
Comment (peterwilsoncc) [reply packages/edit-site/src/components/global-styles/font-library-modal/index.js:56]: Thanks both, I'll leave it as is for now as font-family is using the same permissions.

File: packages/edit-site/src/components/global-styles/font-library-modal/installed-fonts.js [hunk 2]
```diff
@@ -173,11 +175,17 @@ function InstalledFonts() {
 function Footer( { shouldDisplayDeleteButton, handleUninstallClick } ) {
 	const { saveFontFamilies, fontFamiliesHasChanges, isInstalling } =
 		useContext( FontLibraryContext );
+
+	const canUserDelete = useSelect( ( select ) => {
+		const { canUser } = select( coreStore );
+		return canUser( 'delete', 'font-families' );
+	} );
+
```
Comment (peterwilsoncc) [inline packages/edit-site/src/components/global-styles/font-library-modal/installed-fonts.js]: I'm sure it's my mistake but this is always returning false and preventing the button from showing.

**Comment by peterwilsoncc** [comment]
@creativecoder I'm going to have to thinkg about this, I only became aware of the permissions discussion a couple of days ago so still getting my head aound the workings of the controller.

File: packages/edit-site/src/components/global-styles/font-library-modal/installed-fonts.js (hunk 2 above)
Comment (peterwilsoncc) [reply packages/edit-site/src/components/global-styles/font-library-modal/installed-fonts.js]: Yup, this was my mistake. ;)

File: lib/compat/wordpress-6.5/fonts/fonts.php [hunk 3]
```diff
@@ -84,6 +84,32 @@ function gutenberg_create_initial_post_types() {
 	);
 }
 
+/**
+ * Filters the user capabilities to grant the font family capabilities as necessary.
+ *
+ * Files must be modifiable to grant these capabilities and the user must als
```
Comment (getdave) [inline lib/compat/wordpress-6.5/fonts/fonts.php:90]:
```suggestion
 * Files must be modifiable to grant these capabilities and the user must also
```

File: lib/compat/wordpress-6.5/fonts/fonts.php [hunk 4]
```diff
@@ -84,6 +84,32 @@ function gutenberg_create_initial_post_types() {
...
+ * have the `edit_theme_options` capability.
+ *
+ * These are created as faux primitive capabilities to allow for the use
+ * if the delete_post meta capability.
```
Comment (getdave) [inline lib/compat/wordpress-6.5/fonts/fonts.php:94]:
```suggestion
 * of the delete_post meta capability.
```

**Comment by getdave** [comment]
Thanks for all the work on this PR. 

We will need to merge this PR prior to Beta 3 and I'm starting to become concerned that it's now getting very close.

I think we should add some tests, including some Playwright e2e tests to validate that the Upload/Delete options are / aren't available based on certain conditions.

@creativecoder @matiasbenedetto are you able to support with this?

For my part I'm going to manually test the PR and will be around to assist wherever I can.

**Comment by creativecoder** [comment]
@getdave I'm working on a new PR today, combining the approach here with https://github.com/WordPress/gutenberg/pull/59104, based on @peterwilsoncc 's [feedback](https://github.com/WordPress/gutenberg/pull/59104#pullrequestreview-1891687961).

Here's what I'm planning:

- Add new meta cap for font upload, following the same pattern as language packs
- Gating UI install/delete actions based on font family user capabilities

**Comment by peterwilsoncc** [comment]
Closing this in favour of #59294 which includes a combination of this PR and Grant's original PR.

**Pull request closed without merging**